src/node/coverage.ts

```diff
@@ -0,0 +1,267 @@
+    "data:image/svg+xml;base64," +
+      Buffer.from(
+        `<svg version="1.1" width="18" height="18" xmlns="http://www.w3.org/2000/svg">
+  <rect width="3" height="100%" fill="${Colors.missed}" />
+</svg>`
+      ).toString("base64")
+  ),
+} as const;
+
+export function activateCoverage(context: ExtensionContext) {
+  const command = "codecov.reset.api.key";
+
+  const resetApiKeyHandler = () => {
+    context.secrets.delete("api.key");
+    updateDecorations();
+    console.log("reset api key");
+  };
+
+  context.subscriptions.push(
+    commands.registerCommand(command, resetApiKeyHandler)
+  );
+
+  const lineCoveredDecoration = window.createTextEditorDecorationType({
+    gutterIconPath: Icons.covered,
+    overviewRulerColor: Colors.covered,
+    isWholeLine: true,
+    overviewRulerLane: OverviewRulerLane.Right,
+  });
+  const linePartialDecoration = window.createTextEditorDecorationType({
+    gutterIconPath: Icons.partial,
+    overviewRulerColor: Colors.partial,
+    overviewRulerLane: OverviewRulerLane.Right,
+  });
+  const lineMissedDecoration = window.createTextEditorDecorationType({
+    gutterIconPath: Icons.missed,
+    overviewRulerColor: Colors.missed,
+    overviewRulerLane: OverviewRulerLane.Right,
+  });
+
+  let activeEditor = window.activeTextEditor;
+
+  async function updateDecorations() {
+    if (!activeEditor) {
+      return;
+    }
+
+    const config = workspace.getConfiguration("codecov");
+
+    const enabled = config.get("coverage.enabled");
+    if (!enabled) return;
+
+    let apiKey = await context.secrets.get("api.key");
+    let apiUrl = config.get("api.url");
+    const provider = config.get("api.gitProvider");
+
+    if (!apiKey) {
+      const result = await window.showErrorMessage(
+        "To see Codecov line coverage in your editor, you must first set an API Key.",
+        {
+          modal: true,
+          detail:
+            "If you don't want to do this right now, you can disable Codecov line coverage in the Codecov extension's settings.",
+        },
+        "Set an API Key"
+      );
+
+      if (result === "Set an API Key") {
+        apiKey = await window.showInputBox({
+          title: "Enter your Codecov API Key",
+          prompt:
+            "You can generate an API key in your account settings within the Codecov app.",
+        });
+        if (apiKey) await context.secrets.store("api.key", apiKey);
+        updateDecorations();
+        return;
+      }
+    }
+
+    if (!apiUrl) {
+      // Just reset it to default for this workspace
+      await config.update("api.url", "https://api.codecov.io", false);
+      apiUrl = "https://api.codecov.io";
+    }
+
+    const path = encodeURIComponent(
+      workspace.asRelativePath(activeEditor.document.fileName)
+    );
+
+    const pathToWorkspace = workspace.getWorkspaceFolder(
+      Uri.file(activeEditor.document.fileName)
+    )?.uri.path;
+
+    const gitConfig = Uri.file(`${pathToWorkspace}/.git/config`);
```

Comment:
This was a challenge while developing. Vscode provides a git extension, but I couldn't figure out how to get the information I need from it. Alternatively, I found an extension on GitHub that spawns a child process and runs git commands. I settled on reading the necessary info from the `.git` directory as my solution, it seems to do everything I want with minimal overhead.